### ISSUE TITLE
Terragrunt does not respect the Terraform lock file when using dependencies and cache is active

### ISSUE THREAD

**Issue by michelzanini** [gruntwork-io/terragrunt#2509]
It seems that when the Terragrunt file is using dependencies, and Terraform caching is enabled (TF_PLUGIN_CACHE_DIR), when Terragrunt runs initially on the dependencies to obtain their outputs, it doesn't seem to be respecting the lock files inside those dependencies.

I am going to describe several steps to illustrate and reproduce the issue.

In the example we have two folders with terragrunt files as defined here:

```
├── eks
│   ├── .terraform.lock.hcl
│   └── terragrunt.hcl
└── vpc
    ├── .terraform.lock.hcl
    └── terragrunt.hcl
```

Where `vpc` has no dependencies and `eks` depends on the `vpc` as follows:

```
dependency "vpc" {
  config_path = "../vpc"
}
```

Both modules contain a lock file with:

```
provider "registry.terraform.io/hashicorp/aws" {
  version     = "4.60.0"
  constraints = ">= 3.72.0, >= 4.0.0, ~> 4.0"
  hashes = [
    "h1:XxVhnhtrRW3YueabP668hVZ3qL4th7pcWbx+ot/l864=",
    "h1:b2U4EncUaHCsQuiePo/yHZiH5ib0rx+P/qG4LC8pGlA=",
    "h1:dtAnw6tXxtBV4DaASHz1yEONtDn2XptsNJmdzv/X1Gg=",
(...)
```

**Important**: You have to use a lock file with a provider version that is NOT the latest, to illustrate the issue. The day I am writing this `4.61.0` is the latest, but I am using `4.60.0`.

* Define a folder to be the caching folder for provider downloads by defining an env variable for TF_PLUGIN_CACHE_DIR

```
export TF_PLUGIN_CACHE_DIR=~/terraform-cache
```

* Run `terragrunt plan` on `vpc` module for the first time (no .terragrunt-cache present)
* List the cache folder to see the AWS provider:

 ```
ls ~/terraform-cache/registry.terraform.io/hashicorp/aws
> 4.60.0
```

* It correctly downloaded `4.60.0` from the lock file and not `4.61.0`, which is the latest at the moment. **This worked correctly.**
* Now cd into `eks` folder and run `terrragrunt plan` (no .terragrunt-cache present)
* List the cache folder to see the AWS provider:

 ```
ls ~/terraform-cache/registry.terraform.io/hashicorp/aws
> 4.60.0
> 4.61.0
```
* See that `4.61.0` (latest version) shows up even if NONE of my lock files are using it. **This is NOT expected, it should have used the lock files**.

* Debugging this further I discovered that this is downloaded right at the start of the `terragrunt plan` at this initial part:

```
cd eks
terragrunt plan                                                          
WARN[0003] No double-slash (//) found in source URL sandbox/terraform-aws-vpc-modules.git. Relative paths in downloaded Terraform code may not work.  prefix=[/Users/michelzanini/git/sandbox/vpc] 
``` 

* It happens right at the start when executing the `vpc` dependency. 
* The bad thing is the folder version 4.61.0, is never re-used. Every run it will re-download 4.61.0 once more. You can observe by clearing  .terragrunt-cache and running plan again. The provider cache will update the file timestamp date.

**What issues is this causing?**

The biggest issue that we discovered is on CI environment where we can run several jobs in parallel on the same Jenkins slave.
What happens is that `4.61.0` folder is re-downloaded on every run, and when there's too many jobs at the same time, eventually jobs start failing with this error:

```
╷
│ Error: Required plugins are not installed
│ 
│ The installed provider plugins are not consistent with the packages
│ selected in the dependency lock file:
│   - registry.terraform.io/hashicorp/aws: the cached package for registry.terraform.io/hashicorp/aws 4.61.0 (in .terraform/providers) does not match any of the checksums recorded in the dependency lock file
│ 
│ Terraform uses external plugins to integrate with a variety of different
│ infrastructure services. You must install the required plugins before
│ running Terraform operations.
```

Which I believe it might happen because the file gets re-downloaded and the checksum is taken when the file is not fully re-downloaded (since there's several jobs running in the same machine using the same cache folder).

**How it should work?**

Even for dependencies, the lock files should be respected and it should never download `4.61.0`.

**Environment I tested**

```
Terraform 1.4.2
Terragrunt 0.45.0
```

But I also had this issue on:

```
Terraform 1.3.3
Terragrunt 0.39.2
```

**Comment by levkoburburas** [comment]
@michelzanini, could you try if the issue was resolved in [v0.45.12](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.45.12) release?

**Comment by bekahmark12** [comment]
Hello, I am having this exact same issue running on terragrunt version 0.48.1. This has not been resolved and still gives the "required plulgins not installed" error when running on a module with dependencies when using the terragrunt cache. Are there plans to fix this as the current version has not provided a fix?

**Comment by levkohimins** [comment]
Hello @bekahmark12, I tried to repo the issue using **michelzanini**'s description, but I didn't manage, everything worked correctly as expected. Could you please provide your step by step actions to repro this issue?

**Comment by fesplugas-drms** [comment]
Issue is still present on latest Terraform and Terragrunt versions.

Terraform v1.6.4
Terragrunt v0.53.4

**Comment by michelzanini** [comment]
Sorry for taking this long to reply.
I can confirm the issue is fixed with at least Terragrunt `0.50.x` and above.
Thanks.
